### PR TITLE
#161967093 rate an article

### PR DESCRIPTION
**Pull request by denomash**
#### What does this PR do?

- Users should be able to rate and view article rates

#### Description of Task to be completed?

-  a user can rate an article.
- a user can view the average rates of the article.

#### How should this be manually tested?

- Clone this repo from the terminal

```
git clone https://github.com/andela/ah-code-titans-frontend.git
```

- cd into the projects folder
```
cd ah-code-titans-frontend
```

- Then Run `git fetch ft-rate-articles-161967093` 

- checkout the branch: `git checkout ft-rate-articles-161967093`

- Install all dependencies by running: `npm install` or `yarn`

- Run the app by running: `npm start`

#### What are the relevant pivotal tracker stories?
[#161967093](https://www.pivotaltracker.com/story/show/161967093)

**Checklist**

- [x] My code follows the style guidelines of this project

- [x] I have performed a self-review of my own code

- [x] My changes generate no new warnings

####Screenshot
screenshot 2019-01-29 at 18 47 10

### CODE REVIEW

File: src/api/rateArticleApi.js [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+      .post(
+        `/api/article/${slug}/rate`,
+        {
+          article: {
+            rate: rated
+          }
+        },
+        {
+          headers: {
+            Authorization: `Token ${token}`
```
Comment (justmesam) [inline src/api/rateArticleApi.js]: why are we using a hard coded token?

File: src/components/containers/rating/GetRates.js [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+    super(props);
+    this.state = {
+      rating: props.averageRating
+    };
+  }
+
+  componentDidMount() {
+    const { rate } = this.props;
+    const slug = window.location.pathname.slice(9);
+    console.log(slug);
```
Comment (justmesam) [inline src/components/containers/rating/GetRates.js]: no need to push `console.log`

File: src/components/views/SingleArticle.js [hunk 3]
```diff
@@ -0,0 +1,55 @@
+import React from "react";
+import { Container, Divider } from "semantic-ui-react";
+import GetRates from "../containers/rating/GetRates";
+import RateArticle from "../containers/rating/RateArticle";
+
+const SingleArticle = props => (
```
Comment (justmesam) [inline src/components/views/SingleArticle.js:6]: do we need this file?

File: src/reducers/rateArticleReducer.js [hunk 4]
```diff
@@ -0,0 +1,23 @@
+import objectAssign from "object-assign";
+import * as types from "../actions/actionTypes";
+
+export default function rateArticleReducer(state = {}, action = {}) {
+  switch (action.type) {
+    case types.RATE_ARTICLE_SUCCESS: {
+      let rate = objectAssign({}, state.rate);
+      // let rating = objectAssign({}, state.rating);
```
Comment (justmesam) [inline src/reducers/rateArticleReducer.js]: avoid pushing commented code.

**Review by justmesam — CHANGES_REQUESTED** [review]

File: src/components/views/SingleArticle.js (hunk 3 above)
Comment (denomash) [reply src/components/views/SingleArticle.js:6]: we don't need it was temporary to ensure my feature works

File: src/reducers/rateArticleReducer.js (hunk 4 above)
Comment (denomash) [reply src/reducers/rateArticleReducer.js]: let me remove it

File: src/components/containers/rating/GetRates.js (hunk 2 above)
Comment (denomash) [reply src/components/containers/rating/GetRates.js]: let me remove it

File: src/api/rateArticleApi.js (hunk 1 above)
Comment (denomash) [reply src/api/rateArticleApi.js]: This was before integration but I have fixed this.

File: src/api/rateArticleApi.js [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+          }
+        }
+      )
+      .then((response) => {
+        if (response.status === 201) {
+          return { success: true, articleRating: response.data };
+        }
+      })
+      .catch((err) => {
+        throw err;
```
Comment (bettblake08) [inline src/api/rateArticleApi.js]: Please handle the catch clause well. Add a toastr message for example.

File: src/api/rateArticleApi.js [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+  static getArticleRate(slug) {
+    return instance
+      .get(`/api/article/${slug}/rating`)
+      .then((response) => {
+        if (response.status === 200) {
+          return { success: true, articleRating: response.data["Rated at"] };
+        }
+      })
+      .catch((err) => {
+        throw err;
```
Comment (bettblake08) [inline src/api/rateArticleApi.js]: Please handle the catch clause well. Add a toastr message for example.

File: src/assets/style/pages/socialAuthenticationPage.css [hunk 7]
```diff
@@ -0,0 +1,22 @@
+.bg--success, .bg--error {
```
Comment (bettblake08) [inline src/assets/style/pages/socialAuthenticationPage.css]: Please git ignore this css file.

File: src/assets/style/pages/updatePasswordPage.css [hunk 8]
```diff
@@ -0,0 +1,2 @@
+.passwordForm__sendBtn {
```
Comment (bettblake08) [inline src/assets/style/pages/updatePasswordPage.css]: Please git ignore this css file.

File: src/reducers/rateArticleReducer.js [hunk 9]
```diff
@@ -0,0 +1,21 @@
+import objectAssign from "object-assign";
+import * as types from "../actions/actionTypes";
+
+export default function rateArticleReducer(state = {}, action = {}) {
+  switch (action.type) {
+    case types.RATE_ARTICLE_SUCCESS: {
+      let rate = objectAssign({}, state.rate);
```
Comment (bettblake08) [inline src/reducers/rateArticleReducer.js]: This can be simplified to this `return { ...state, rating: action.data };`

**Review by bettblake08 — CHANGES_REQUESTED** [review]

File: src/actions/__test__/rateArticleActions.test.js [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+import * as types from "../actionTypes";
+import * as actionCreators from "../rateArticleActions";
+
+const middlewares = [thunk];
+const mockStore = configureMockStore(middlewares);
+
+// test action creators
+describe("get average rates action", () => {
+  it("should return a get action", () => {
+    const store = mockStore({});
```
Comment (Darius-Ndubi) [inline src/actions/__test__/rateArticleActions.test.js]: You can have the constants set before the test suite to avoid repetition.

Comment (denomash) [reply src/actions/__test__/rateArticleActions.test.js]: True. Let me work on it

File: src/assets/style/pages/updatePasswordPage.css (hunk 8 above)
Comment (denomash) [reply src/assets/style/pages/updatePasswordPage.css]: Is it not necessary

File: src/api/rateArticleApi.js (hunk 6 above)
Comment (denomash) [reply src/api/rateArticleApi.js]: Cool. on it

**Comment by denomash** [comment]
@bettblake08 @Darius-Ndubi I have worked on your feedback. Kindly review my work again for any other change.

**Review by jillo-abdullahi — CHANGES_REQUESTED** [review]
Hi @denomash, nicely done. Here is one further input from me. When the user is logged out and tries to view the article, they can still try to rate it but the error that pops up pertains to comments. We can put it so that the rate component doesn't show up when the user isn't logged in. Please check the screenshot below.

screenshot 2019-01-29 at 10 44 50

**Comment by denomash** [comment]
@jillo-abdullahi made the change

**Review by davishooly — APPROVED** [review]

**Review by bettblake08 — APPROVED** [review]

File: src/api/rateArticleApi.js [hunk 11]
```diff
@@ -0,0 +1,59 @@
...
+        }
+      )
+      .then((response) => {
+        if (response.status === 201) {
+          return { success: true, articleRating: response.data };
+        }
+      })
+      .catch((err) => {
+        if (err.response.status === 400) {
+          toastr.error("AN error occured try again");
```
Comment (Bryan-Cee) [inline src/api/rateArticleApi.js:22]: Fix the Typo in `AN` to be `An`

File: src/assets/style/articles/style.css [hunk 12]
```diff
@@ -0,0 +1,115 @@
+html {
```
Comment (Bryan-Cee) [inline src/assets/style/articles/style.css]: Please remove this `css` file, we are supposed to write the css in our `scss` files

**Review by Darius-Ndubi — APPROVED** [review]
👍

File: src/reducers/index.js [hunk 13]
```diff
@@ -5,12 +5,14 @@ import registrationReducer from "./registrationReducer";
...
 
 export default history => combineReducers({
   router: connectRouter(history),
   registrationReducer,
   loginReducer,
   userSettingsReducer,
   profileReducer,
-  article
+  article,
+  rateArticleReducer
```
Comment (Bryan-Cee) [inline src/reducers/index.js]: Please remove the suffix, `reducer` in your reducer to be `rateArticle`

**Review by Bryan-Cee — APPROVED** [review]

**Review by jillo-abdullahi — APPROVED** [review]

File: src/reducers/index.js (hunk 13 above)
Comment (denomash) [reply src/reducers/index.js]: fixed it

File: src/assets/style/articles/style.css (hunk 12 above)
Comment (denomash) [reply src/assets/style/articles/style.css]: removed it

File: src/api/rateArticleApi.js (hunk 11 above)
Comment (denomash) [reply src/api/rateArticleApi.js:22]: fixed the typo

File: src/store/initialState.js [hunk 14]
```diff
@@ -45,5 +45,9 @@ export default {
     tag: {
       articles: {}
     }
+  },
+  rateArticleReducer: {
```
Comment (justmesam) [inline src/store/initialState.js]: remove the reducer suffix

File: src/reducers/index.js [hunk 15]
```diff
@@ -16,5 +17,6 @@ export default history => combineReducers({
   profileReducer,
   article,
   tag,
-  commentReducer
+  commentReducer,
```
Comment (justmesam) [inline src/reducers/index.js]: remove the reducer suffix

File: src/store/initialState.js (hunk 14 above)
Comment (denomash) [reply src/store/initialState.js]: On it

File: src/components/containers/rating/GetRates.js (hunk 2 above)
Comment (denomash) [reply src/components/containers/rating/GetRates.js]: I have fixed it

File: src/reducers/index.js (hunk 15 above)
Comment (denomash) [reply src/reducers/index.js]: I have fixed it

File: src/store/initialState.js (hunk 14 above)
Comment (denomash) [reply src/store/initialState.js]: I have fixed it

**Pull request merged**